nanocoap: add coap_opt_add_uquery2() with parameter key value length

pokgak opened the pull request:
### Contribution description

This PR introduces a variant of the function `coap_opt_add_uquery()` to nanocoap which also accepts requires value length and key length. The original function assumes the key and value are null-terminated, which may not be the case if the values are located in the same buffer, separated by a separator as discussed in https://github.com/RIOT-OS/RIOT/pull/10222#discussion_r358257934.

### Testing procedure

Run the nanocoap unittests from RIOT root: `make -C tests/unittests tests-nanocoap make all term`

It should not affect existing usages of `coap_opt_add_uquery()`.

smlng commented on sys/net/application_layer/gcoap/gcoap.c:
```diff
@@ -990,23 +990,30 @@ ssize_t gcoap_encode_link(const coap_resource_t *resource, char *buf,
 
 int gcoap_add_qstring(coap_pkt_t *pdu, const char *key, const char *val)
 {
+    return gcoap_add_qstring2(pdu, key, strlen(key), val, strlen(val));
```
this is not safe and will (or can?) crash when `val = NULL` as `strlen(NULL)` is no good. Hence this should be changed to

```
    size_t val_len = (val) ? strlen(val) : 0;
    return gcoap_add_qstring2(pdu, key, strlen(key), val, val_len);
```

smlng replied:
also you should really check `key` here, too -- for the same reason. So add `assert(key);` at the beginning should suffice.

kb2ma replied:
Yes, the gcoap unit tests include a query with zero length value, and it segfaults as is.

smlng commented on sys/net/application_layer/gcoap/gcoap.c:
```diff
@@ -990,23 +990,30 @@ ssize_t gcoap_encode_link(const coap_resource_t *resource, char *buf,
+{
+    assert(pdu);
+    assert(key);
+    assert(key_len != 0);
```
nit-picky: why not simply `assert(key_len)`? I mean otherwise you should also do `assert(pdu != NULL)` and `assert(key != NULL)`, or not 😉

pokgak commented:
@smlng @kb2ma sorry for the late response.

I rebased to master and reworked the function to extend `coap_opt_add_uquery()` instead. I also addressed comments from @smlng's review.

kb2ma commented on sys/include/net/nanocoap.h:
```diff
@@ -987,6 +990,24 @@ ssize_t coap_opt_add_opaque(coap_pkt_t *pkt, uint16_t optnum, const uint8_t *val
+ * @param[in]     val         Value to assign to @p key (may be NULL)
+ * @param[in]     val_len     Length of @p val. 0 if @p val is NULL
+ *
+ * @pre     ((pkt != NULL) && (key != NULL))
```
Also add `key_len > 0` since that assert is in the code.

pokgak replied:
added `key_len > 0`

kb2ma commented on sys/net/application_layer/nanocoap/nanocoap.c:
```diff
@@ -848,25 +848,40 @@ ssize_t coap_opt_add_string(coap_pkt_t *pkt, uint16_t optnum, const char *string
 
 ssize_t coap_opt_add_uquery(coap_pkt_t *pdu, const char *key, const char *val)
```
Use `pkt` rather than `pdu` here for consistency with the rest of nanocoap. I felt "pdu" was more descriptive and so used that in gcoap.

pokgak replied:
changed to `pkt`.

kb2ma commented on sys/net/application_layer/nanocoap/nanocoap.c:
```diff
@@ -848,25 +848,40 @@ ssize_t coap_opt_add_string(coap_pkt_t *pkt, uint16_t optnum, const char *string
-    /* test if the query string fits, account for the zero termination */
-    if ((len + val_len + 1) >= NANOCOAP_QS_MAX) {
+    /* account for '=' */
+    if (val) {
```
It would be more compact vertically to use:
```
val_len = val ? (val_len + 1) : 0;
```

pokgak replied:
changed to one line.

kb2ma commented on sys/net/application_layer/nanocoap/nanocoap.c:
```diff
@@ -848,25 +848,40 @@ ssize_t coap_opt_add_string(coap_pkt_t *pkt, uint16_t optnum, const char *string
+    }
+
+    /* test if the query string fits, account for the zero termination
+     * Account for '\0'*/
```
Don't need to say it twice. ;-)

pokgak replied:
oops :)

kb2ma commented on sys/net/application_layer/nanocoap/nanocoap.c:
```diff
@@ -848,25 +848,40 @@ ssize_t coap_opt_add_string(coap_pkt_t *pkt, uint16_t optnum, const char *string
-        memcpy(&qs[len + 1], val, (val_len - 1));
-        len += val_len;
+    memcpy(&qs[0], key, key_len);
+    if (val && val_len > 0) {
```
Either just use `(val && val_len)` or add the extra parens `(val && (val_len > 0))`

pokgak replied:
changed to `(val && val_len)`.

kb2ma commented on sys/net/application_layer/nanocoap/nanocoap.c:
```diff
@@ -846,27 +846,36 @@ ssize_t coap_opt_add_string(coap_pkt_t *pkt, uint16_t optnum, const char *string
 
     /* test if the query string fits, account for the zero termination */
-    if ((len + val_len + 1) >= NANOCOAP_QS_MAX) {
+    if ((key_len + val_len + 1) >= NANOCOAP_QS_MAX) {
```
Just noticed -- only need `>` comparison, not `>=`.

kb2ma commented on sys/net/application_layer/nanocoap/nanocoap.c:
```diff
@@ -846,27 +846,36 @@ ssize_t coap_opt_add_string(coap_pkt_t *pkt, uint16_t optnum, const char *string
-    size_t val_len = (val) ? (strlen(val) + 1) : 0;
+
+    /* account for '=' */
+    val_len = val ? (val_len + 1) : 0;
```
Need to check both `val` and `val_len` before incrementing:

```
val_len = (val && val_len) ? (val_len + 1) : 0;
```

kb2ma commented:
Fixups look good. 

When I looked through the code again, I tried a little experiment and created the line below just after defining `qs`.
```
    size_t qs_len = key_len + ((val && val_len) ? 1 : 0) + val_len + 1;
```
I found that using it made the function noticeably clearer by not redefining `val_len`. It also had no impact on ROM or RAM. I'm OK with the function as is, but I'd be happy to review incorporation of this change.

pokgak commented:
Nice idea! I also find changing val_len to be a bit confusing. Incorporated your suggestion in the latest fixup.

kb2ma commented:
Let's give @smlng some time to comment, otherwise I will approve.

smlng commented on sys/net/application_layer/nanocoap/nanocoap.c:
```diff
@@ -846,27 +846,34 @@ ssize_t coap_opt_add_string(coap_pkt_t *pkt, uint16_t optnum, const char *string
+    assert(pkt);
+    assert(key);
+    assert(key_len);
+
```
you could add an assert for `val` here too, like
```
assert(!val || (val && val_len))
```

smlng replied:
better say: you should add this assert, see other comments

pokgak replied:
Using the test from https://github.com/RIOT-OS/RIOT/pull/13022#discussion_r386330125, I think the following would be better:

```C
assert((!val || !val_len) || (val && val_len));
```

kb2ma replied:
My turn. :-) I think the assert should be
```
assert(!val_len || (val && val_len))
```
I think `!val && val_len` is a programming error. OTOH `val && !val_len` says to ignore the value, which is odd but valid.

smlng commented on sys/net/application_layer/nanocoap/nanocoap.c:
```diff
@@ -846,27 +846,34 @@ ssize_t coap_opt_add_string(coap_pkt_t *pkt, uint16_t optnum, const char *string
     char qs[NANOCOAP_QS_MAX];
-    size_t len = strlen(key);
-    size_t val_len = (val) ? (strlen(val) + 1) : 0;
+    size_t qs_len = key_len + ((val && val_len) ? 1 : 0) + val_len + 1;
```
this looks overly complicated to me, I'd suggest:
```
size_t  qs_len = key_len + (val ? (val_len + 1) : 0);
```

the extra `+1` is not needed, see below

smlng replied:
the `assert` above helps here too, bc without the assert one could pass `val = NULL`  and `val_len = 42`, which might not be good

pokgak replied:
yep that is simpler

pokgak replied:
changed to the following to ignore `val_len` if `val` is `NULL`:
```C
size_t qs_len = key_len + ((val && val_len) ? (val_len + 1) : 0);
```

kb2ma replied:
I agree with this 100%! It is simpler more idiomatic to avoid the terminator in these calculations.

smlng commented on sys/net/application_layer/nanocoap/nanocoap.c:
```diff
@@ -846,27 +846,34 @@ ssize_t coap_opt_add_string(coap_pkt_t *pkt, uint16_t optnum, const char *string
 
     /* test if the query string fits, account for the zero termination */
-    if ((len + val_len + 1) >= NANOCOAP_QS_MAX) {
+    if (qs_len > NANOCOAP_QS_MAX) {
```
change this to `>=` bc no extra `+1` above

pokgak replied:
changed to `>=`

smlng commented on sys/net/application_layer/nanocoap/nanocoap.c:
```diff
@@ -846,27 +846,34 @@ ssize_t coap_opt_add_string(coap_pkt_t *pkt, uint16_t optnum, const char *string
+        memcpy(&qs[key_len + 1], val, val_len);
     }
-    qs[len] = '\0';
+    qs[qs_len - 1] = '\0';
```
change to `qs[qs_len]` due to change above

pokgak replied:
changed

kb2ma commented:
Tested, and works for me. Nice discussion and improvements. Given all the discussion here, I think a unit test would be valuable. I added a [gist](https://gist.github.com/kb2ma/7ea2a541289a8f7b1bb17b4924ec36d9) as a starting point. It assumes asserts are turned off for unit tests.

smlng commented:
please squash and trigger ci

kb2ma commented on sys/net/application_layer/nanocoap/nanocoap.c:
```diff
@@ -846,27 +846,35 @@ ssize_t coap_opt_add_string(coap_pkt_t *pkt, uint16_t optnum, const char *string
+    assert(pkt);
+    assert(key);
+    assert(key_len);
+    assert((!val || !val_len) || (val && val_len));
```
This line is not correct. It is always returns true. Should be:

```
assert(!val_len || (val && val_len))
```

smlng replied:
Maybe we could be *very* explicit here and do `assert ((!val && !val_len) || (val && val_len))` to clearly state we want either both or none, as `val = NULL`  with `val_len > 0` would likely be an error, too.

kb2ma replied:
I guess it comes down to the purpose of this assert. The other three asserts notify the developer of an unquestionable runtime problem before a crash. I think this assert should only do the same. I am concerned about prohibiting a developer from doing something when we don't have the context to know if it is a mistake (val && !val_len).

pokgak replied:
changed to `assert(!val_len || (val && val_len))`

kb2ma commented:
> Tested, and works for me.

Sorry for the misleading statement. See the comment above about the assert. i also would like to see a unit test based on the gist.

smlng commented:
please fix and squash directly, so we get this merged

kb2ma commented:
I am OK with leaving off the unit test for this PR. I am happy to add it afterwards so we can make progress. :-)

pokgak commented:
Squashed and rebased. I'll work on the test in another PR :)
